integrations/jenkins/client.py

```diff
@@ -0,0 +1,110 @@
+    ) -> None:
+        self.jenkins_base_url = jenkins_base_url
+        self.client = http_async_client
+        self.client.auth = httpx.BasicAuth(jenkins_user, jenkins_token)
+
+    async def get_all_jobs_and_builds(
+        self,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        cache_key = "jenkins_jobs_and_builds"
+
+        if cache := event.attributes.get(cache_key):
+            logger.info("picking from cache")
+            yield cache
+            return
+
+        async for jobs in self.fetch_jobs_and_builds_from_api():
+            event.attributes.setdefault(cache_key, []).extend(jobs)
+            yield jobs
+
+    async def fetch_jobs_and_builds_from_api(
+        self, parent_job: Optional[dict[str, Any]] = None
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        page_size = 5
+        page = 0
+
+        while True:
+            params = self._build_api_params(page_size, page)
+            base_url = self._build_base_url(parent_job)
+
+            job_response = await self.client.get(f"{base_url}/api/json", params=params)
+            job_response.raise_for_status()
+            jobs = job_response.json()["jobs"]
+
+            if not jobs:
+                break
+
+            yield await self._process_jobs(jobs, parent_job)
+
+            page += 1
+
+            if len(jobs) < page_size:
+                break
+
+    def _build_api_params(self, page_size: int, page: int) -> dict[str, Any]:
+        start_idx = page_size * page
+        end_idx = start_idx + page_size
+
+        return {
+            "tree": f"jobs[name,url,displayName,fullName,color,jobs[name,color,fullName,displayName,url],"
+            f"builds[id,number,url,result,duration,timestamp,displayName,fullDisplayName]"
+            f"{{0,50}}]{{{start_idx},{end_idx}}}"
+        }
+
+    def _build_base_url(self, parent_job: Optional[dict[str, Any]]) -> str:
+        job_path = urlparse(parent_job["url"]).path if parent_job else ""
+        return f"{self.jenkins_base_url}{urlparse(job_path).path}"
+
+    async def _process_jobs(
+        self, jobs: list[dict[str, Any]], parent_job: Optional[dict[str, Any]]
+    ) -> list[dict[str, Any]]:
+        """
+        Process the list of jobs, optionally attaching information from the parent job.
+        If a job has no builds, recursively fetch and include its child jobs.
+
+        Args:
+            jobs (list[dict]): List of jobs to process.
+            parent_job (Optional[dict]): Parent job information to attach to each job.
+
+        Returns:
+            list[dict]: Processed list of jobs with optional parent job information.
+        """
+        job_batch = []
+
+        for job in jobs:
+            if parent_job:
+                job["__parentJob"] = parent_job
+
+            if "builds" not in job:
+                async for child_jobs in self.fetch_jobs_and_builds_from_api(job):
+                    job_batch.extend(child_jobs)
```

Comment:
Maybe u should gather all sub jobs in a list and then go over them after yielding the first jobs? Because right now i will have to wait for the recursive to finish before ingesting anything

Reply:
We now immediately extract the buildable jobs (orange circle) on a page, followed by the child jobs (green). We then proceed to the next page and repeat the process.

<img width="1168" alt="Screenshot 2024-01-10 at 14 22 46" src="https://github.com/port-labs/ocean/assets/8656262/33d65b6b-aaac-472b-8cb9-a494d63c389d">

Reply:
Think about this scenario
If you have one root job that holds all the jobs underneath it you will paginate over the jenkins jobs and get one job and then you will see its a folder job and query all its children jobs but you wont start to yield information out which will cause all the jobs in page the your got to wait for this jobs to process all children jobs recursively


what im expecting that will happen is:
- fetch page
- save the child jobs in a list
- yield jobs
- .... continue paging
- after paging go over the child jobs and run the same logic in a recursive way

very important to make sure that any information is not waiting for child information to return

Reply:
I have now changed to this pattern